File: ios/Classes/BridgingCreator.swift
import Flutter
import SuperwallKit
import UIKit

/// Plugin state enum to track attachment status
enum PluginState {
  case attached
  case detached
  case attaching
}

/// Creates a method channel for a particular unique instance of a class
public class BridgingCreator: NSObject, FlutterPlugin {
  static private var _shared: BridgingCreator? = nil
  static var shared: BridgingCreator {
    guard let shared = _shared else {
      fatalError(
        "Attempting to access the shared BridgingCreator before `register(with registrar: FlutterPluginRegistrar)` has been called."
      )
    }

    return shared
  }
  
  // Add state tracking
  private static var pluginState: PluginState = .detached
  
  // Keep reference to last registrar for reattachment
  private static var lastRegistrar: FlutterPluginRegistrar?
  
  // Check if plugin is healthy
  static func checkPluginHealth() -> Bool {
    return pluginState == .attached && _shared != nil
  }
  
  // Force plugin reattachment
  static func forceReattachment() -> Bool {
    if pluginState == .attached {
      return true // Already attached
    }
    
    // Set state to attaching to prevent concurrent attempts
    pluginState = .attaching
    
    guard let registrar = lastRegistrar else {
      pluginState = .detached
      print("ERROR: No registrar available for reattachment")
      return false
    }
    
    do {
      let communicator = Communicator(
        name: "SWK_BridgingCreator", binaryMessenger: registrar.messenger())
      
      let bridge = BridgingCreator(registrar: registrar, communicator: communicator)
      BridgingCreator._shared = bridge
      
      registrar.addMethodCallDelegate(bridge, channel: communicator)
      pluginState = .attached
      print("Successfully reattached Superwall plugin")
      return true
    } catch {
      pluginState = .detached
      print("Error during plugin reattachment: \(error)")
      return false
    }
  }

  let registrar: FlutterPluginRegistrar

  private let communicator: Communicator
  private var instances: [String: BridgeInstance] = [:]

  init(registrar: FlutterPluginRegistrar, communicator: Communicator) {
    self.registrar = registrar
    self.communicator = communicator
    super.init()
  }

  // Enhanced bridgeInstance with better error handling and retry
  func bridgeInstance(for bridgeInstance: BridgeId) -> Any? {
    // Try to retrieve the instance
    if let instance = instances[bridgeInstance] {
      return instance
    }
    
    // If not found, check if we're detached and try to reattach
    if BridgingCreator.pluginState != .attached {
      if BridgingCreator.forceReattachment() {
        // Try again after reattachment
        return instances[bridgeInstance]
      }
    }
    
    // Still not found, provide better error message
    assertionFailure("Unable to find a bridge instance for \(bridgeInstance) after reattachment attempt.")
    return nil
  }
  
  // Type-specific convenience method
  func bridgeInstanceTyped<T>(for bridgeInstance: BridgeId) -> T? {
    if let instance = self.bridgeInstance(for: bridgeInstance) as? T {
      return instance
    }
    return nil
  }

  public static func register(with registrar: FlutterPluginRegistrar) {
    // Store registrar for later reattachment if needed
    lastRegistrar = registrar
    
    let communicator = Communicator(
      name: "SWK_BridgingCreator", binaryMessenger: registrar.messenger())

    let bridge = BridgingCreator(registrar: registrar, communicator: communicator)
    BridgingCreator._shared = bridge
    
    // Set plugin state to attached
    pluginState = .attached
    print("Superwall plugin successfully attached")

    registrar.addMethodCallDelegate(bridge, channel: communicator)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "createBridgeInstance":
      guard
        let bridgeId: String = call.argument(for: "bridgeId")
      else {
        print("WARNING: Unable to create bridge instance")
        result(FlutterError(code: "BAD_ARGS", message: "Missing bridgeId", details: nil))
        return
      }

      let initializationArgs: [String: Any]? = call.argument(for: "args")

      do {
        createBridgeInstance(bridgeId: bridgeId, initializationArgs: initializationArgs)
        result(nil)
      } catch {
        result(FlutterError(code: "CREATION_FAILED", message: "Bridge creation failed: \(error)", details: nil))
      }
    
    case "checkPluginHealth":
      // New method to allow Dart side to check plugin health
      result(BridgingCreator.checkPluginHealth())
      
    case "forceReattachment":
      // New method to allow Dart side to force reattachment
      result(BridgingCreator.forceReattachment())

    default:
      result(FlutterMethodNotImplemented)
    }
  }

  // Create the bridge instance as instructed from Dart
  @discardableResult
  private func createBridgeInstance(
    bridgeId: BridgeId, initializationArgs: [String: Any]? = nil
  ) -> BridgeInstance {
    // An existing bridge instance might exist if it were created natively, instead of from Dart
    if let existingBridgeInstance = instances[bridgeId] {
      return existingBridgeInstance
    }

    guard let bridgeClass = BridgingCreator.Constants.bridgeMap[bridgeId.bridgeClass] else {
      assertionFailure(
        "Unable to find a bridgeClass for \(bridgeId.bridgeClass). Make sure to add to BridgingCreator+Constants.swift"
      )
      return BridgeInstance(bridgeId: bridgeId, initializationArgs: initializationArgs)
    }

    let bridgeInstance = bridgeClass.init(
      bridgeId: bridgeId, initializationArgs: initializationArgs)
    instances.updateValue(bridgeInstance, forKey: bridgeId)

    registrar.addMethodCallDelegate(bridgeInstance, channel: bridgeInstance.communicator)
    bridgeInstance.events()
    return bridgeInstance
  }

  // Create the bridge instance as instructed from native
  func createBridgeInstance(bridgeClass: BridgeClass, initializationArgs: [String: Any]? = nil)
    -> BridgeInstance
  {
    return createBridgeInstance(
      bridgeId: bridgeClass.generateBridgeId(), initializationArgs: initializationArgs)
  }
}

File: android/src/main/kotlin/com/superwall/superwallkit_flutter/BridgingCreator.kt
package com.superwall.superwallkit_flutter

import android.os.Handler
import android.os.Looper
import android.util.Log
import com.superwall.superwallkit_flutter.bridges.BridgeClass
import com.superwall.superwallkit_flutter.bridges.BridgeId
import com.superwall.superwallkit_flutter.bridges.BridgeInstance
import com.superwall.superwallkit_flutter.bridges.Communicator
import com.superwall.superwallkit_flutter.bridges.bridgeClass
import com.superwall.superwallkit_flutter.bridges.generateBridgeId
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.flow.filterNotNull
import java.util.concurrent.ConcurrentHashMap
import io.flutter.embedding.android.FlutterActivity
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.first
import io.flutter.plugin.common.EventChannel
import java.util.concurrent.atomic.AtomicReference

// Plugin state enum to track attachment status
enum class PluginState {
    ATTACHED, DETACHED, ATTACHING
}

class BridgingCreator(
    val flutterPluginBinding: suspend () -> FlutterPlugin.FlutterPluginBinding,
    val scope: CoroutineScope = CoroutineScope(Dispatchers.IO)
) : MethodCallHandler {
    private val instances: MutableMap<String, BridgeInstance> = ConcurrentHashMap()

    object Constants {}

    companion object {
        private val lock = Any()
        private var _shared: MutableStateFlow<BridgingCreator?> = MutableStateFlow(null)
        
        // Add plugin state tracking
        private val pluginState = AtomicReference<PluginState>(PluginState.DETACHED)
        
        // Static reference to last known binding
        @Volatile
        var lastKnownBinding: FlutterPlugin.FlutterPluginBinding? = null
            private set

        // Main thread handler for UI operations
        private val mainHandler = Handler(Looper.getMainLooper())
            
        suspend fun shared() : BridgingCreator = _shared.filterNotNull().first()

        private var _flutterPluginBinding: MutableStateFlow<FlutterPlugin.FlutterPluginBinding?> =
            MutableStateFlow(null)

        suspend fun waitForPlugin() : FlutterPlugin.FlutterPluginBinding {
            return _flutterPluginBinding.filterNotNull().first()
        }
        
        // Add method to force reattachment
        fun forceReattachment(binding: FlutterPlugin.FlutterPluginBinding?) {
            val currentState = pluginState.get()
            if (currentState != PluginState.ATTACHED) {
                // Set state to attaching to prevent concurrent attempts
                if (!pluginState.compareAndSet(currentState, PluginState.ATTACHING)) {
                    // Another thread is already attaching
                    return
                }
                
                synchronized(lock) {
                    try {
                        // Use last known binding if available
                        val bindingToUse = binding ?: lastKnownBinding
                        if (bindingToUse == null) {
                            pluginState.set(PluginState.DETACHED) // Reset state
                            println("ERROR: No binding available for reattachment")
                            return
                        }
                        
                        // Execute on main thread to avoid potential threading issues
                        mainHandler.post {
                            try {
                                // Create new bridge instance
                                val bridge = BridgingCreator({ bindingToUse })
                                _shared.value = bridge
                                _flutterPluginBinding.value = bindingToUse
                                val communicator = Communicator(bindingToUse.binaryMessenger, "SWK_BridgingCreator")
                                communicator.setMethodCallHandler(bridge)
                                
                                pluginState.set(PluginState.ATTACHED)
                                println("Successfully reattached Superwall plugin")
                            } catch (e: Exception) {
                                pluginState.set(PluginState.DETACHED) // Reset state on error
                                println("Error during plugin reattachment on main thread: ${e.message}")
                                e.printStackTrace()
                            }
                        }
                    } catch (e: Exception) {
                        pluginState.set(PluginState.DETACHED) // Reset state on error
                        println("Error during plugin reattachment: ${e.message}")
                        e.printStackTrace()
                    }
                }
            }
        }
        
        // Add method to check plugin health
        fun checkPluginHealth(): Boolean {
            return pluginState.get() == PluginState.ATTACHED && _shared.value != null
        }
        
        fun setFlutterPlugin(binding: FlutterPlugin.FlutterPluginBinding) {
            // Save last known binding first (even if we don't use it now)
            lastKnownBinding = binding
            
            // Only allow binding to occur once unless we're deliberately reattaching
            if (_flutterPluginBinding.value != null && pluginState.get() == PluginState.ATTACHED) {
                println("WARNING: Attempting to set a flutter plugin binding again. Current state: ${pluginState.get()}")
                return
            }

            binding?.let {
                synchronized(lock) {
                    val bridge = BridgingCreator({ waitForPlugin() })
                    _shared.value = bridge
                    _flutterPluginBinding.value = binding
                    val communicator = Communicator(binding.binaryMessenger, "SWK_BridgingCreator")
                    communicator.setMethodCallHandler(bridge)
                    
                    // Update plugin state to attached
                    pluginState.set(PluginState.ATTACHED)
                    println("Superwall plugin successfully attached")
                }
            }
        }
    }

    // Enhanced tearDown that tracks state but preserves instance
    fun tearDown() {
        // Mark as detached but maintain references for later reattachment
        pluginState.set(PluginState.DETACHED)
        println("BridgingCreator tearDown called - maintaining instance for reattachment")
        // Do NOT set _shared.value = null
        // Do NOT set _flutterPluginBinding.value = null
    }

    // Enhanced bridgeInstance with retry logic
    suspend fun <T> bridgeInstance(bridgeId: BridgeId): T? {
        var attempts = 0
        val maxAttempts = 3
        
        while (attempts < maxAttempts) {
            try {
                BreadCrumbs.append("BridgingCreator.kt: Searching for $bridgeId among ${instances.count()}: ${instances.toFormattedString()}")
                var instance = instances[bridgeId] as? T
                
                if (instance == null) {
                    if (attempts < maxAttempts - 1) {
                        // Try reattachment before final attempt
                        val binding = SuperwallkitFlutterPlugin.instance?.let { 
                            _flutterPluginBinding.value ?: lastKnownBinding 
                        }
                        forceReattachment(binding)
                        delay(300L * (attempts + 1)) // Exponential backoff
                        attempts++
                        continue
                    }
                    throw AssertionError("Unable to find a native instance for $bridgeId after $attempts attempts. Logs: ${BreadCrumbs.logs()}")
                }
                
                return instance
            } catch (e: Exception) {
                if (attempts < maxAttempts - 1) {
                    attempts++
                    delay(300L * (attempts + 1)) // Exponential backoff
                    continue
                }
                throw e
            }
        }
        
        throw AssertionError("Maximum retry attempts exceeded for $bridgeId")
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        scope.launch {
            try {
                when (call.method) {
                    "createBridgeInstance" -> {
                        val bridgeId = call.argument<String>("bridgeId")
                        val initializationArgs = call.argument<Map<String, Any>>("args")
                        if (bridgeId != null) {
                            createBridgeInstanceFromBridgeId(bridgeId, initializationArgs)
                            result.success(null)
                        } else {
                            println("WARNING: Unable to create bridge")
                            result.badArgs(call)
                        }
                    }
                    "checkPluginHealth" -> {
                        // New method to allow Dart side to check plugin health
                        result.success(checkPluginHealth())
                    }
                    "forceReattachment" -> {
                        // New method to allow Dart side to force reattachment
                        val binding = _flutterPluginBinding.value ?: lastKnownBinding
                        forceReattachment(binding)
                        result.success(true)
                    }
                    else -> result.notImplemented()
                }
            } catch (e: Exception) {
                println("Error in BridgingCreator.onMethodCall: ${e.message}")
                e.printStackTrace()
                result.error("EXCEPTION", e.message, null)
            }
        }
    }

    // Create the bridge instance as instructed from Dart
    private suspend fun createBridgeInstanceFromBridgeId(
        bridgeId: BridgeId,
        initializationArgs: Map<String, Any>?
    ): BridgeInstance {
        val existingBridgeInstance = instances[bridgeId]

        existingBridgeInstance?.let {
            if(existingBridgeInstance.cachable)
                return it
        }

        val bridgeInstance = bridgeInitializers[bridgeId.bridgeClass()]?.invoke(
            flutterPluginBinding().applicationContext,
            bridgeId,
            initializationArgs
        )

        bridgeInstance?.let { bridgeInstance ->
            instances[bridgeId] = bridgeInstance
            bridgeInstance.communicator().setMethodCallHandler(bridgeInstance)
            bridgeInstance.events()
            return bridgeInstance
        } ?: run {
            throw AssertionError("Unable to find a bridge initializer for ${bridgeId}. Make sure to add to BridgingCreator+Constants.kt.}")
        }
    }

    // Create the bridge instance as instructed from native
    suspend fun createBridgeInstanceFromBridgeClass(
        bridgeClass: BridgeClass,
        initializationArgs: Map<String, Any>? = null
    ): BridgeInstance {
        return createBridgeInstanceFromBridgeId(bridgeClass.generateBridgeId(), initializationArgs)
    }
}